_posts/2019/05/2019-05-24-checking-driven-development.md

```diff
@@ -0,0 +1,150 @@
+    I think the difference is building products for repeated use
+    versus combining those products and their own code to get one specific result.
+
+4.  So how do scientists figure out if their software is doing the right thing?
+    The answer is spot checks:
+    each time they produce an intermediate or final result,
+    they scan a table, create a chart, or inspect some summary statistics
+    to see if everything looks OK.
+    Their heuristics are usually easy to state,
+    like "there shouldn't be NAs at this point" or "the age range should be reasonable",
+    but applying those heuristics to a particular analysis always depends on
+    the data scientist's evolving insight into the data in question.
+
+I therefore agree that
+there are more important things to teach to novice data scientists than unit testing.
+(I can already hear my software engineering friends muttering, "Unclean! Unclean!")
+What I think we *should* teach is defensive programming:
+the checks that data scientists do while they're doing analyses
+should be recorded in their code as assertions.
+This will help reusability---it's amazing how often a one-off analysis
+winds up being used many times---but the real goal is comprehensibility.
+If I can get your code and data,
+run the former on the latter,
+and get the same result that you did,
+then your computation is reproducible,
+but that doesn't mean I can understand it.
+Prose explanations help,
+which is part of why knuths<sup><a href="#footnote-knuth">1</a></sup> are so popular,
+but comments and paragraphs won't check that assumptions and invariants still hold
+when the person trying to reproduce the analysis starts fiddling around with it.
+And unlike comments,
+runnable assertions can't fall out of step with what the code is actually doing...
+
+By analogy with [test-driven development](https://en.wikipedia.org/wiki/Test-driven_development),
+we could call this process "checking-driven development".
+Just like writing tests forces you to be explicit about what "done" looks like,
+adding checks to analyses forces you to be explicit about what "right" looks like.
+I believe this approach is pedagogically defensible:
+
+1.  Knowing how to write useful assertions
+    is necessary (but not sufficient) for writing unit tests.
+2.  It's the-same-as-but-better-than what good practitioners currently do.
+
+## An Example
+
+So what might this look like in practice?
+Imagine we have this simple dplyr pipeline
+to find the mean age of people who aren't "alpha":
+
+```
+data <- tribble(
+  ~record_id, ~person_id, ~age,
+  100,         "alpha",    17,
+  200,         "alpha",    34,
+  300,         "beta",     21,
+  400,         "gamma",    NA,
+  500,         "gamma",    26
+)
+
+data %>%
+  filter(person_id != "alpha") %>%
+  group_by(person_id) %>%
+  summarize(midpoint = mean(age))
+```
+
+Now imagine a trio of functions `nickr_row`, `nickr_col`, and `nickr_group`
```

Comment:
Not sure how much overlap there is with nickr, but there is already a CRAN package for doing data checking using the [assertr](https://cran.r-project.org/web/packages/assertr/vignettes/assertr.html) package. It's quite full featured.

For more defensive programming in functions there's the [assertive](https://cran.r-project.org/web/packages/assertive/index.html) package.

Reply:
Yeah, I looked at `assertr` last Nov/Dec, but struggled to make sense of `assert` vs `verify` vs `insist`. It also doesn't integrate with logging libraries, so there's no way to say "send messages *here* in development but *there* in production".